File: 12-valid-parentheses/solution.java
// Valid Parentheses
// Time: O(n), Space: O(n)

import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put(')', '(');
        mapping.put('}', '{');
        mapping.put(']', '[');
        
        for (char c : s.toCharArray()) {
            if (mapping.containsKey(c)) {
                char top = stack.isEmpty() ? '#' : stack.pop();
                if (mapping.get(c) != top) {
                    return false;
                }
            } else {
                stack.push(c);
            }
        }
        
        return stack.isEmpty();
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.isValid("()[]{}"));   // true
        System.out.println(sol.isValid("([)]"));     // false
        System.out.println(sol.isValid("{[]}"));     // true
    }
}

File: 12-valid-parentheses/solution.go
// Valid Parentheses
// Time: O(n), Space: O(n)

package main

import "fmt"

type Solution struct{}

func (s *Solution) IsValid(str string) bool {
	stack := []rune{}
	mapping := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}
	
	for _, c := range str {
		if match, exists := mapping[c]; exists {
			var top rune
			if len(stack) > 0 {
				top = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
			} else {
				top = '#'
			}
			if match != top {
				return false
			}
		} else {
			stack = append(stack, c)
		}
	}
	
	return len(stack) == 0
}

func main() {
	sol := &Solution{}
	fmt.Println(sol.IsValid("()[]{}"))   // true
	fmt.Println(sol.IsValid("([)]"))     // false
	fmt.Println(sol.IsValid("{[]}"))     // true
}
